### ISSUE TITLE
force thread/process/pool by kernel

### ISSUE THREAD

**Issue by jonathan-beard** [RaftLib/RaftLib#21]
In the RaftLib C++Now tutorial (full slide deck here: http://www.jonathanbeard.io/pdf/cppnow2016.pdf) I proposed this syntax for partitioning a VM space:
![screen shot 2016-08-22 at 11 32 53 am](https://cloud.githubusercontent.com/assets/1173230/17862746/32e3a0d8-685c-11e6-964f-6c81db131128.png)

Seems intuitive at first, however when you have multiple ports on a single compute kernel then you begin to have issues. So I'm looking at something that works more like this:

```cpp
raft::map m;

raft::kernel a,c;
/** all of b forced to a single process, call acts as a decorator **/
auto b_proc( raft::parallel::process( b ) );
m += a >> b_proc >> c;
m.exe();
```

Any thoughts?

**Comment by adishavit** [comment]
If you do not need to modify `b_proc`, would it be possible to write this:
```
raft::map m;

raft::kernel a,b,c;
/** all of b forced to a single process, call acts as a decorator **/
m += a >> raft::parallel::process( b ) >> c;
m.exe();
```
I'm kinda out of my depth here, but another possible syntax would be an iomanip style manipulator:
```
raft::map m;

raft::kernel a,b,c;
/** all of b forced to a single process, call acts as a decorator **/
m += a >> raft::parallel::process >> b >> c;
m.exe();
```
Both suggestions, obviously, assume you do not need access to the temporary object created by `raft::parallel::process` and that the downstream operator knows how to take and handle (e.g. copy or move from it) an rvalue (which will evaporate after the statement is done).

**Comment by jonathan-beard** [comment]
I actually like the iomanip style best, however if you have multiple incoming ports to "b" then it would seem you have to reiterate on ever stream...or perhaps one is sufficient...seems a bit awkward either way....when awkward is what I'm trying to avoid in the first place.

**Comment by adishavit** [comment]
What would a multi-input b code look like?
Can you give some examples of this case in various syntices (!?) ?

**Comment by jonathan-beard** [comment]
**Single port in-line iomanip-like syntax:**
```cpp
raft::map m;
raft::kernel a,b,c;
/** all of b forced to a single process, call acts as a decorator **/
m += a >> raft::parallel::process >> b >> c;
m.exe();
```

**Dual port in-line iomanip-like syntax:**
```cpp
raft::map m;
raft::kernel a,c,d;
/** fake dual port kernel for example purposes **/
raft::dualPortKernel b;
/** all of b forced to a single process, call acts as a decorator **/
m += a >> raft::parallel::process >> b[ "x0" ] >> c;
m += d >> raft::parallel::process >> b[ "x1" ];
m.exe();
```

Visually this would look something like this:
<img width="288" alt="graph" src="https://cloud.githubusercontent.com/assets/1173230/18060741/714f0392-6de5-11e6-9d8c-b66a25c73c12.png">

However, using the decorator syntax from above, we could shorten to this:
```cpp
raft::map m;
raft::kernel a,c,d;
/** all of b forced to a single process, call acts as a decorator **/
auto b_proc( raft::parallel::process( b ) );
m += a >> b_proc[ "x0" ] >> c;
m += d >> b_proc[ "x1" ];
m.exe();
```

Again, I don't anticipate that this type of syntax would be needed that often. I'm just concerned/dismayed that the solution I like the most with the iomanip-like operators requires the user to specify  raft::parallel::process twice (and remember to do so). I suppose we could throw an exception if that weren't the case, however that is still clunky. 

I'm definitely open to anybody with better solutions..but it seems for a (hopefully) little used feature, that decorators as the last example shows, might be the best option in the end. That being said, you could also have a modifier on the decorator constructor to also force core affinity or hardware type.

**Comment by adishavit** [comment]
I agree that the second option does indeed seem more readable in the case of multiple ports. 
Is there something preventing both usages?
The iomanipulator might be named slightly differently so there is no confusion. 
Just like std manipulator a that are named differently than the corresponding stream methods.
